File: 백준/Silver/9012. 괄호/괄호.py
TESTCASE = int(input())

for _ in range(TESTCASE):
    stack = []
    string = input()
    correct_sign = True
    for ch in string:
        if ch == '(':
            stack.append(ch)
        else:
            if len(stack) == 0:
                correct_sign = False
                break
            stack.pop()

    if len(stack) > 0 or not correct_sign:
        print("NO")
    else:
        print("YES")

File: 백준/Silver/9012. 괄호/괄호.cs
using System;
using System.Collections.Generic;


class Program
{
    public static void Main(string[] args)
    {
        string a = Console.ReadLine();
        int n = int.Parse(a);

        for (int i = 0; i < n; i++)
        {
            string  b = Console.ReadLine();
            Stack<char> stack = new Stack<char>();
            bool q = true;
            for (int j = 0; j < b.Length; j++)
            {
                char c = b[j];
                if(c=='(')
                {
                    stack.Push(c);
                }
                else
                {
                    if (stack.Count == 0)
                    {
                        q = false;
                        break;
                    }
                    if (stack.Peek() == '(')
                    {
                        stack.Pop();
                    }
                }

            }

            if (!q || stack.Count > 0)
            {
                Console.WriteLine("NO");
                continue;
            }
            else Console.WriteLine("YES");
            
        }
    }
}